core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -579,39 +598,46 @@ class KafkaApis(val requestChannel: RequestChannel,
             // record the bytes out metrics only when the response is being sent
             brokerTopicStats.updateBytesOut(topicPartition.topic, fetchRequest.isFromFollower, data.records.sizeInBytes)
           }
           response
         }
-
-        trace(s"Sending Fetch response with partitions.size=${unconvertedFetchResponse.responseData().size()}, " +
-          s"metadata=${unconvertedFetchResponse.sessionId()}")
-
-        if (fetchRequest.isFromFollower)
-          sendResponseExemptThrottle(request, createResponse(0))
-        else
-          sendResponseMaybeThrottle(request, requestThrottleMs => createResponse(requestThrottleMs))
       }
 
-      // When this callback is triggered, the remote API call has completed.
-      // Record time before any byte-rate throttling.
-      request.apiRemoteCompleteTimeNanos = time.nanoseconds
-
       if (fetchRequest.isFromFollower) {
         // We've already evaluated against the quota and are good to go. Just need to record it now.
         val responseSize = sizeOfThrottledPartitions(versionId, unconvertedFetchResponse, quotas.leader)
         quotas.leader.record(responseSize)
-        fetchResponseCallback(bandwidthThrottleTimeMs = 0)
+        trace(s"Sending Fetch response with partitions.size=${unconvertedFetchResponse.responseData().size()}, " +
+          s"metadata=${unconvertedFetchResponse.sessionId()}")
+        sendResponseExemptThrottle(request, createResponse(0))
       } else {
         // Fetch size used to determine throttle time is calculated before any down conversions.
         // This may be slightly different from the actual response size. But since down conversions
         // result in data being loaded into memory, it is better to do this after throttling to avoid OOM.
         val responseStruct = unconvertedFetchResponse.toStruct(versionId)
-        quotas.fetch.maybeRecordAndThrottle(request.session.sanitizedUser, clientId, responseStruct.sizeOf,
-          fetchResponseCallback)
+
+        trace(s"Sending Fetch response with partitions.size=${unconvertedFetchResponse.responseData().size()}, " +
+          s"metadata=${unconvertedFetchResponse.sessionId()}")
+
+        // Record both bandwidth and request quota-specific values and throttle by muting the channel if any of the quotas
+        // have been violated. If both quotas have been violated, use the max throttle time between the two quotas.
+        val bandwidthThrottleTimeMs = quotas.fetch.maybeRecord(request, responseStruct.sizeOf)
+        val requestThrottleTimeMs = quotas.request.maybeRecord(request)
+
+        val maxThrottleTimeMs = bandwidthThrottleTimeMs.max(requestThrottleTimeMs)
+        val throttledChannel: Option[ThrottledChannel] =
+          if (maxThrottleTimeMs > 0) {
+            if (bandwidthThrottleTimeMs > requestThrottleTimeMs)
+              quotas.fetch.maybeThrottle(request, bandwidthThrottleTimeMs)
+            else quotas.request.maybeThrottle(request, requestThrottleTimeMs)
```

Comment:
It maybe more readable to keep the code style consistent by moving `quotas.request.maybeThrottle(request, requestThrottleTimeMs)` to a new line with `{` and `}`.

Reply:
Done